File: backends/aspnet-core/Dtos/CapsuleDtos.cs
using System.ComponentModel.DataAnnotations;

namespace HelloTime.AspNetCore.Dtos;

public sealed class CreateCapsuleRequest
{
    [Required(ErrorMessage = "title 不能为空")]
    [MaxLength(100, ErrorMessage = "title 最多 100 个字符")]
    public string Title { get; init; } = string.Empty;

    [Required(ErrorMessage = "content 不能为空")]
    public string Content { get; init; } = string.Empty;

    [Required(ErrorMessage = "creator 不能为空")]
    [MaxLength(30, ErrorMessage = "creator 最多 30 个字符")]
    public string Creator { get; init; } = string.Empty;

    [Required(ErrorMessage = "openAt 不能为空")]
    public string OpenAt { get; init; } = string.Empty;
}

public sealed record CapsuleCreatedResponse(
    string Code,
    string Title,
    string Creator,
    string OpenAt,
    string CreatedAt
);

public sealed record CapsuleDetailResponse(
    string Code,
    string Title,
    string? Content,
    string Creator,
    string OpenAt,
    string CreatedAt,
    bool Opened
);

public sealed record CapsulePageResponse(
    IReadOnlyList<CapsuleDetailResponse> Content,
    int TotalElements,
    int TotalPages,
    int Number,
    int Size
);

File: backends/vapor/server/Sources/App/DTOs/CapsuleDTOs.swift
import Vapor

struct CreateCapsuleRequest: Content, Validatable {
    let title: String
    let content: String
    let creator: String
    let openAt: String

    static func validations(_ validations: inout Validations) {
        validations.add("title", as: String.self, is: .count(1...100), required: true)
        validations.add("content", as: String.self, is: !.empty, required: true)
        validations.add("creator", as: String.self, is: .count(1...30), required: true)
        validations.add("openAt", as: String.self, is: !.empty, required: true)
    }
}

struct CapsuleCreatedResponse: Content {
    let code: String
    let title: String
    let creator: String
    let openAt: String
    let createdAt: String
}

struct CapsuleDetailResponse: Content {
    let code: String
    let title: String
    let content: String?
    let creator: String
    let openAt: String
    let createdAt: String
    let opened: Bool
}

struct CapsulePageResponse: Content {
    let content: [CapsuleDetailResponse]
    let totalElements: Int
    let totalPages: Int
    let number: Int
    let size: Int
}

struct HealthStatusResponse: Content {
    let status: String
    let timestamp: String
    let techStack: TechStackResponse
}

struct TechStackResponse: Content {
    let framework: String
    let language: String
    let database: String
}
